bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/GardenaSmartImpl.java

```diff
@@ -12,562 +12,377 @@
-                int status = responseException.getResponse().getStatus();
-                if (status == 401) {
-                    throw new GardenaUnauthorizedException(ex.getCause());
-                }
-            }
-            throw new GardenaException(ex.getMessage(), ex);
-        } catch (Exception ex) {
-            throw new GardenaException(ex.getMessage(), ex);
+            httpClient.stop();
+        } catch (Exception e) {
+            // ignore
         }
+        httpClient.destroy();
+        locationsResponse = new LocationsResponse();
+        allDevicesById.clear();
+        initialized = false;
     }
 
     /**
-     * Verifies the Gardena Smart Home session and reconnects if necessary.
+     * Restarts all websockets.
      */
-    private void verifySession() throws GardenaException {
-        if (session == null
-                || session.getCreated() + (config.getSessionTimeout() * 60000) <= System.currentTimeMillis()) {
-            logger.trace("(Re)logging in to Gardena Smart Home");
-            session = executeRequest(HttpMethod.POST, URL_LOGIN, new GardenaConfigWrapper(config), SessionWrapper.class)
-                    .getSession();
+    @Override
+    public synchronized void restartWebsockets() {
+        logger.debug("Restarting GardenaSmart Webservice");
+        stopWebsockets();
+        try {
+            startWebsockets();
+        } catch (Exception ex) {
+            logger.warn("Restarting GardenaSmart Webservice failed: {}, restarting binding", ex.getMessage());
+            eventListener.onError();
         }
     }
 
     /**
-     * Thread which refreshes the data from Gardena Smart Home.
+     * Sets the dataItem from the websocket event into the correct device.
      */
-    private class RefreshDevicesThread implements Runnable {
-        private boolean connectionLost = false;
-
-        @Override
-        public void run() {
-            try {
-                logger.debug("Refreshing gardena device data");
-                final Map<String, Device> newDevicesById = new HashMap<>();
-
-                for (Location location : allLocations) {
-                    Devices devices = loadDevices(location);
-                    for (Device device : devices.getDevices()) {
-                        if (DEVICE_CATEGORY_GATEWAY.equals(device.getCategory())) {
-                            location.getDeviceIds().remove(device.getId());
-                        } else {
-                            newDevicesById.put(device.getId(), device);
+    private void handleDataItem(final DataItem<?> dataItem) throws GardenaException {
+        final String deviceId = dataItem.getDeviceId();
+        Device device = allDevicesById.get(deviceId);
+        if (device == null && !(dataItem instanceof LocationDataItem)) {
+            device = new Device(deviceId);
+            allDevicesById.put(device.id, device);
+
+            if (initialized) {
+                scheduler.schedule(() -> {
```

Comment:
You have to keep track of the future returned here so you can cancel it in case the handler is disposed.